### PR TITLE
Emit LLVM bitcode without using LLVM

### LINKED ISSUE #13265: directly output LLVM bitcode rather than using LLVM's IRBuilder API
Zig can be built with or without `-Denable-llvm`. Currently, Zig is not very useful without enabling LLVM extensions. However, as we move into the future, Zig is intending to compete directly with LLVM, making builds of Zig without LLVM a compelling option for the backends directly supported by Zig.

There are a few reasons why one might want an LLVM-less binary:
 * The executable is 4.4 MiB instead of 169 MiB.
 * Bootstrapping it only requires a C compiler instead of requiring a modern C++ compiler, Python 3, bash, and CMake (also C++).
   - This would make it much easier to obtain a Zig compiler on a new operating system or a limited environment such as a calculator.

This proposal is to treat LLVM bitcode files (.bc) as the target output format, rather than going through the C++ IRBuilder API. This would make it possible for even non-LLVM-enabled builds of Zig to still output LLVM IR that could be consumed by Clang, other LLVM tools, or integrated with other software.

One example user story comes from [Roc](https://www.roc-lang.org/). I'd like to get @rtfeldman's take on this - I know that you're using Zig to output .bc files, but then what happens? Does a different tool compile that together with other code, or do you use Zig for the final link step too? I'm guessing that Roc would be able to use the non-LLVM-enabled Zig binaries for their use case.

There is a second major reason for this proposal, which is perhaps even the better argument in favor of it, which is to make incremental compilation work more robustly. As the Zig project moves forward, we want to make `CacheMode.incremental` the default for all backends including LLVM (caddbbc315e834241c6ae7e22434b574c589e5fc). This means we would want to save the LLVM IR module (.bc) with every compilation and restore it for subsequent compilations, using the IRBuilder API to add and remove declarations as necessary from the LLVM IR module, keeping the .bc file on disk in sync for future incremental compilations.

However... the API lacks functionality. For example, aliases cannot be deleted:

https://github.com/ziglang/zig/blob/e67c756b9114debdaa566188f677f60e507dfac8/src/codegen/llvm.zig#L1330-L1348

If Zig were in control of outputting the .bc file instead, then Zig could simply not emit aliases that are not supposed to exist. We no longer are limited by what the IRBuilder API can do. This would make the LLVM backend very similar to the WebAssembly backend in the sense that it gains a linking component and directly outputs the module.

Finally, in the incremental compilation sense, Zig would already be trying to keep a .bc file on disk up-to-date via the IRBuilder API. Doing it directly instead of via a limited API is a more direct way to solve the problem, and the performance would be in our hands rather than in the hands of the LLVM project.

I think these two reasons combined make this proposal worth seriously considering, despite the downsides of taking on additional maintenance with LLVM upgrades, and introducing an entirely new class of bugs from generating malformed .bc files.

### PR DESCRIPTION
**Pull request by antlilja**
This PR should close #13265

With this PR bitcode is created through the `Builder.zig`, this bitcode is then parsed into a module by LLVM through `LLVMParseBitcodeInContext2` and then compiled into object files as before. The LLVM `DIBuilder` has been replaced by a system in `Builder.zig`. This PR also removes all uses of the LLVM library inside `Builder.zig` and removes a lot of the bindings in `bindings.zig` and `zig_llvm.h/.cpp`.

## Performance
Hare are some `perf` stats of a release safe version of the compiler compiling itself (on the llvm-bc branch) with an empty cache:
### llvm-bc branch (ReleaseSafe):
```
Performance counter stats for 'zig build -Dstatic-llvm --zig-lib-dir lib --search-prefix zig-bootstrap/release/x86_64-linux-musl-native -Dno-langref -Dno-autodocs -Dtarget=x86_64-linux-musl -p stage3 install':

        209,102.81 msec task-clock:u                     #    2.236 CPUs utilized
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
         8,347,566      page-faults:u                    #   39.921 K/sec
   787,734,595,025      cycles:u                         #    3.767 GHz
     3,345,576,458      stalled-cycles-frontend:u        #    0.42% frontend cycles idle
    10,671,110,379      stalled-cycles-backend:u         #    1.35% backend cycles idle
   970,481,802,496      instructions:u                   #    1.23  insn per cycle
                                                  #    0.01  stalled cycles per insn
   195,371,479,830      branches:u                       #  934.332 M/sec
     6,637,680,252      branch-misses:u                  #    3.40% of all branches

      93.498796731 seconds time elapsed

     172.877783000 seconds user
      36.207457000 seconds sys
```
### master branch (ReleaseSafe):
```
 Performance counter stats for 'zig build -Dstatic-llvm --zig-lib-dir lib --search-prefix zig-bootstrap/release/x86_64-linux-musl-native -Dno-langref -Dno-autodocs -Dtarget=x86_64-linux-musl -p stage3 install':

        220,426.12 msec task-clock:u                     #    2.026 CPUs utilized
                 0      context-switches:u               #    0.000 /sec
                 0      cpu-migrations:u                 #    0.000 /sec
         9,726,043      page-faults:u                    #   44.124 K/sec
   821,802,723,056      cycles:u                         #    3.728 GHz
     3,671,828,684      stalled-cycles-frontend:u        #    0.45% frontend cycles idle
     9,380,976,638      stalled-cycles-backend:u         #    1.14% backend cycles idle
 1,025,703,403,709      instructions:u                   #    1.25  insn per cycle
                                                  #    0.01  stalled cycles per insn
   206,035,478,692      branches:u                       #  934.714 M/sec
     6,999,582,116      branch-misses:u                  #    3.40% of all branches

     108.792765074 seconds time elapsed

     179.451902000 seconds user
      40.951977000 seconds sys
```

### CODE REVIEW

**Comment by antlilja** [comment]
Thanks for the cleanup changes @jacobly0! 

I saw your changes to how debug locations are dealt with, I implemented it the same way before, but I realized that debug locations would be emitted twice, once as `Location` metadata and once as a `DEBUG_LOC`. I did not realize that this would cause a problem when emitting text IR. I think LLVM just creates metadata locations from `DEBUG_LOC` records. But I think we could solve it by just pretend emitting all debug locations after all other metadata. It would be a bit more convoluted inside the text IR emitting code but would probably not be too bad and it is probably the less hot path. I think this could reduce the amount bitcode we have to emit by quite a lot but I would have to measure.

**Comment by jacobly0** [comment]
Don't worry, I noticed the extra bitcode that would be emitted, and the plan is certainly to not do that, but I'm still working on just trying to get things working.  I think there are a few different things to try, and I'll have a better idea after messing with the code more.

**Comment by jacobly0** [comment]
> It seems like I've fixed the problems with getting debug locations working now

Thanks! With that, hello world LLVM IR output now verifies with my commit:
```
$ zig build-exe hello.zig --verbose-llvm-ir=hello.ll -fno-emit-bin
$ opt -p verify hello.ll -o /dev/null
$
```

File: src/codegen/llvm.zig [hunk 1]
```diff
@@ -1967,7 +1970,6 @@ pub const Object = struct {
                 if (!ty.hasRuntimeBitsIgnoreComptime(mod)) {
                     const debug_enum_type = try o.makeEmptyNamespaceDebugType(owner_decl_index);
                     try o.debug_type_map.put(gpa, ty, debug_enum_type);
-                    try o.debug_enums.append(gpa, debug_enum_type);
```
Comment (antlilja) [inline src/codegen/llvm.zig]: Did you mean to remove this line?

Comment (jacobly0) [reply src/codegen/llvm.zig]: Yes, llvm [errors](https://github.com/llvm/llvm-project/blob/6009708b4367171ccdbf4b5905cb6a803753fe18/llvm/lib/IR/Verifier.cpp#L1356-L1363) with `invalid enum type` if it contains anything other than `!DICompositeType(tag: DW_TAG_enumeration_type, ...)`.

**Comment by antlilja** [comment]
> The only remaining important missing feature I can find right now is debug metadata attached to globals:
> 
> ```
> $ cat repro.zig
> export const foo: i32 = 42;
> $ zig build-obj repro.zig --verbose-llvm-ir=repro.ll --verbose-llvm-bc=repro.bc
> $ grep @foo repro.ll
> @foo = dso_local constant i32 42, align 4, !dbg !0
> $ llvm-dis repro.bc -o - | grep @foo
> @foo = dso_local constant i32 42, align 4
> ```

Done!

I also have an idea for a new debug location system which should have a less clunky interface than my last one without over emitting debug locations. And a plan for how to deal with these "phantom" debug locations when emitting text ir. I'll ping you when it's done.

File: src/codegen/llvm/Builder.zig [hunk 2]
```diff
@@ -14296,6 +14292,51 @@ pub fn toBitcode(self: *Builder, allocator: Allocator) bitcode_writer.Error![]co
                 });
             }
 
+            // Write global attached metadata
+            {
+                for (self.variables.items) |variable| {
```
Comment (jacobly0) [inline src/codegen/llvm/Builder.zig:14297]: Is there a reason to not just iterate `self.globals` here?

**Review by jacobly0 — APPROVED** [review]
I now consider this ready to land, as it seems to have reached feature parity with the old implementation.  Any potential improvements can always be done in a future PR as needed.

File: src/codegen/llvm/Builder.zig (hunk 2 above)
Comment (antlilja) [reply src/codegen/llvm/Builder.zig:14297]: Not really, I could change that. The only reason I do them separately is that function metadata should for declarations this way (for definitions it's emitted inside the function block). But I could just as well check for functions specifically.

**Comment by jacobly0** [comment]
> I also have an idea for a new debug location system which should have a less clunky interface than my last one without over emitting debug locations. And a plan for how to deal with these "phantom" debug locations when emitting text ir. I'll ping you when it's done.

I have a fix for this, but it can emit the same debug location multiple times in LLVM IR, so I'll wait to see what your plan is.

**Comment by antlilja** [comment]
> Nice work, @antlilja! I'm looking forward to merging this.
> 
> By the way, did you take any peak RSS measurements with these changes? I'm curious to know how it affected memory usage.

I haven't but I'll definitely take a look at that as well when benchmarking some improvements to debug locations.

**Comment by jacobly0** [comment]
I realize these aren't completely comparable, but I compared the compiler compiling itself to unoptimized bitcode with an empty cache from before the llvm rewrite started to after this change lands:

3bada8e3ce9ba72f57c6fbed100c76fd40ba0d15: (before rewrite)
13.08s
2.214172 GB RSS

edb6486b3bf7a1c333d7cc3348f88ab121b72830: (after rewrite)
19.49s
2.836292 GB RSS

edb6486b3bf7a1c333d7cc3348f88ab121b72830 + the following patch:
13.87s
1.0562 GB RSS

```diff
diff --git a/src/codegen/llvm.zig b/src/codegen/llvm.zig
index 5ea749d6d9..4eee469cc6 100644
--- a/src/codegen/llvm.zig
+++ b/src/codegen/llvm.zig
@@ -1187,14 +1187,8 @@ pub const Object = struct {
             }
         }
 
-        var bitcode_arena_allocator = std.heap.ArenaAllocator.init(
-            std.heap.page_allocator,
-        );
-        errdefer bitcode_arena_allocator.deinit();
-
-        const bitcode = try self.builder.toBitcode(
-            bitcode_arena_allocator.allocator(),
-        );
+        const bitcode = try self.builder.toBitcode(self.gpa);
+        defer self.gpa.free(bitcode);
 
         if (options.pre_bc_path) |path| {
             var file = try std.fs.cwd().createFile(path, .{});
@@ -1250,7 +1244,6 @@ pub const Object = struct {
 
             break :blk module;
         };
-        bitcode_arena_allocator.deinit();
 
         const target_triple_sentinel =
             try self.gpa.dupeZ(u8, self.builder.target_triple.slice(&self.builder).?);
diff --git a/src/codegen/llvm/Builder.zig b/src/codegen/llvm/Builder.zig
index a5aeb7dee3..38903c6289 100644
--- a/src/codegen/llvm/Builder.zig
+++ b/src/codegen/llvm/Builder.zig
@@ -14945,7 +14945,7 @@ pub fn toBitcode(self: *Builder, allocator: Allocator) bitcode_writer.Error![]co
         try strtab_block.end();
     }
 
-    return bitcode.toSlice();
+    return bitcode.toOwnedSlice();
 }
 
 const Allocator = std.mem.Allocator;
diff --git a/src/codegen/llvm/bitcode_writer.zig b/src/codegen/llvm/bitcode_writer.zig
index bfb406d087..d48a92dd40 100644
--- a/src/codegen/llvm/bitcode_writer.zig
+++ b/src/codegen/llvm/bitcode_writer.zig
@@ -40,9 +40,9 @@ pub fn BitcodeWriter(comptime types: []const type) type {
             self.buffer.deinit();
         }
 
-        pub fn toSlice(self: BcWriter) []const u32 {
+        pub fn toOwnedSlice(self: *BcWriter) Error![]const u32 {
             std.debug.assert(self.bit_count == 0);
-            return self.buffer.items;
+            return self.buffer.toOwnedSlice();
         }
 
         pub fn length(self: BcWriter) usize {
```

**Comment by antlilja** [comment]
> I realize these aren't completely comparable, but I compared the compiler compiling itself to unoptimized bitcode with an empty cache from before the llvm rewrite started to after this change lands:
> 
> [3bada8e](https://github.com/ziglang/zig/commit/3bada8e3ce9ba72f57c6fbed100c76fd40ba0d15): (before rewrite) 13.08s 2.214172 GB RSS
> 
> [edb6486](https://github.com/ziglang/zig/commit/edb6486b3bf7a1c333d7cc3348f88ab121b72830): (after rewrite) 19.49s 2.836292 GB RSS
> 
> [edb6486](https://github.com/ziglang/zig/commit/edb6486b3bf7a1c333d7cc3348f88ab121b72830) + the following patch: 13.87s 1.0562 GB RSS
> 
> ```diff
> diff --git a/src/codegen/llvm.zig b/src/codegen/llvm.zig
> index 5ea749d6d9..4eee469cc6 100644
> --- a/src/codegen/llvm.zig
> +++ b/src/codegen/llvm.zig
> @@ -1187,14 +1187,8 @@ pub const Object = struct {
>              }
>          }
>  
> -        var bitcode_arena_allocator = std.heap.ArenaAllocator.init(
> -            std.heap.page_allocator,
> -        );
> -        errdefer bitcode_arena_allocator.deinit();
> -
> -        const bitcode = try self.builder.toBitcode(
> -            bitcode_arena_allocator.allocator(),
> -        );
> +        const bitcode = try self.builder.toBitcode(self.gpa);
> +        defer self.gpa.free(bitcode);
>  
>          if (options.pre_bc_path) |path| {
>              var file = try std.fs.cwd().createFile(path, .{});
> @@ -1250,7 +1244,6 @@ pub const Object = struct {
>  
>              break :blk module;
>          };
> -        bitcode_arena_allocator.deinit();
>  
>          const target_triple_sentinel =
>              try self.gpa.dupeZ(u8, self.builder.target_triple.slice(&self.builder).?);
> diff --git a/src/codegen/llvm/Builder.zig b/src/codegen/llvm/Builder.zig
> index a5aeb7dee3..38903c6289 100644
> --- a/src/codegen/llvm/Builder.zig
> +++ b/src/codegen/llvm/Builder.zig
> @@ -14945,7 +14945,7 @@ pub fn toBitcode(self: *Builder, allocator: Allocator) bitcode_writer.Error![]co
>          try strtab_block.end();
>      }
>  
> -    return bitcode.toSlice();
> +    return bitcode.toOwnedSlice();
>  }
>  
>  const Allocator = std.mem.Allocator;
> diff --git a/src/codegen/llvm/bitcode_writer.zig b/src/codegen/llvm/bitcode_writer.zig
> index bfb406d087..d48a92dd40 100644
> --- a/src/codegen/llvm/bitcode_writer.zig
> +++ b/src/codegen/llvm/bitcode_writer.zig
> @@ -40,9 +40,9 @@ pub fn BitcodeWriter(comptime types: []const type) type {
>              self.buffer.deinit();
>          }
>  
> -        pub fn toSlice(self: BcWriter) []const u32 {
> +        pub fn toOwnedSlice(self: *BcWriter) Error![]const u32 {
>              std.debug.assert(self.bit_count == 0);
> -            return self.buffer.items;
> +            return self.buffer.toOwnedSlice();
>          }
>  
>          pub fn length(self: BcWriter) usize {
> ```

Yeah now that I'm thinking about it it's much more reasonable to not use an arena as it will definitely over allocate.

**Comment by andrewrk** [comment]
Here is my measurement of the impact of this change:

```
Benchmark 1 (3 runs): master/zig build-exe ...
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          85.7s  ± 2.98s     82.2s  … 87.7s           0 ( 0%)        0%
  peak_rss           4.71GB ±  763KB    4.71GB … 4.71GB          0 ( 0%)        0%
  cpu_cycles          330G  ± 2.28G      328G  …  332G           0 ( 0%)        0%
  instructions        460G  ±  303M      460G  …  460G           0 ( 0%)        0%
  cache_references   23.4G  ±  184M     23.3G  … 23.6G           0 ( 0%)        0%
  cache_misses       2.01G  ±  206M     1.78G  … 2.19G           0 ( 0%)        0%
  branch_misses      2.22G  ± 2.09M     2.22G  … 2.23G           0 ( 0%)        0%
Benchmark 2 (3 runs): this-pr/zig build-exe ...
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           109s  ± 2.25s      106s  …  110s           0 ( 0%)        💩+ 26.9% ±  7.0%
  peak_rss           4.92GB ±  544KB    4.92GB … 4.92GB          0 ( 0%)        💩+  4.5% ±  0.0%
  cpu_cycles          409G  ± 3.65G      405G  …  412G           0 ( 0%)        💩+ 23.7% ±  2.1%
  instructions        592G  ±  283M      592G  …  592G           0 ( 0%)        💩+ 28.7% ±  0.1%
  cache_references   26.1G  ± 30.6M     26.1G  … 26.1G           0 ( 0%)        💩+ 11.4% ±  1.3%
  cache_misses       3.53G  ±  141M     3.38G  … 3.66G           0 ( 0%)        💩+ 75.3% ± 19.9%
  branch_misses      2.54G  ± 1.52M     2.53G  … 2.54G           0 ( 0%)        💩+ 14.0% ±  0.2%
```

**Comment by antlilja** [comment]
> Here is my measurement of the impact of this change building the self-hosted compiler:
> 
> ```
> Benchmark 1 (3 runs): master/zig build-exe ...
>   measurement          mean ± σ            min … max           outliers         delta
>   wall_time          85.7s  ± 2.98s     82.2s  … 87.7s           0 ( 0%)        0%
>   peak_rss           4.71GB ±  763KB    4.71GB … 4.71GB          0 ( 0%)        0%
>   cpu_cycles          330G  ± 2.28G      328G  …  332G           0 ( 0%)        0%
>   instructions        460G  ±  303M      460G  …  460G           0 ( 0%)        0%
>   cache_references   23.4G  ±  184M     23.3G  … 23.6G           0 ( 0%)        0%
>   cache_misses       2.01G  ±  206M     1.78G  … 2.19G           0 ( 0%)        0%
>   branch_misses      2.22G  ± 2.09M     2.22G  … 2.23G           0 ( 0%)        0%
> Benchmark 2 (3 runs): this-pr/zig build-exe ...
>   measurement          mean ± σ            min … max           outliers         delta
>   wall_time           109s  ± 2.25s      106s  …  110s           0 ( 0%)        💩+ 26.9% ±  7.0%
>   peak_rss           4.92GB ±  544KB    4.92GB … 4.92GB          0 ( 0%)        💩+  4.5% ±  0.0%
>   cpu_cycles          409G  ± 3.65G      405G  …  412G           0 ( 0%)        💩+ 23.7% ±  2.1%
>   instructions        592G  ±  283M      592G  …  592G           0 ( 0%)        💩+ 28.7% ±  0.1%
>   cache_references   26.1G  ± 30.6M     26.1G  … 26.1G           0 ( 0%)        💩+ 11.4% ±  1.3%
>   cache_misses       3.53G  ±  141M     3.38G  … 3.66G           0 ( 0%)        💩+ 75.3% ± 19.9%
>   branch_misses      2.54G  ± 1.52M     2.53G  … 2.54G           0 ( 0%)        💩+ 14.0% ±  0.2%
> ```
> 
> @antlilja did you perhaps get your branches mixed up when measuring?

Yeah it's definitely possible I missed or did something weird when switching between branches. Seems like the most reasonable explanation, there haven't been any changes since the benchmark that should have that huge of an impact. The only major thing that is different is how were emitting debug locations but is definitely not the cause of that difference. My bad.

**Comment by andrewrk** [comment]
Well, I'm just glad we caught it now. Thanks for all your efforts on this branch.

Here's a new data point that includes #19069 vs old master:

```
Benchmark 1 (3 runs): master/zig build-exe ...
  measurement          mean ± σ            min … max           outliers         delta
  wall_time          84.6s  ± 2.89s     81.3s  … 86.7s           0 ( 0%)        0%
  peak_rss           4.68GB ± 70.2MB    4.64GB … 4.76GB          0 ( 0%)        0%
  cpu_cycles          330G  ± 6.41G      327G  …  338G           0 ( 0%)        0%
  instructions        464G  ± 12.6G      456G  …  478G           0 ( 0%)        0%
  cache_references   23.9G  ±  427M     23.6G  … 24.4G           0 ( 0%)        0%
  cache_misses       1.78G  ±  239M     1.51G  … 1.93G           0 ( 0%)        0%
  branch_misses      2.25G  ± 66.6M     2.21G  … 2.32G           0 ( 0%)        0%
Benchmark 2 (3 runs): llvm-bc+19069/zig build-exe ...
  measurement          mean ± σ            min … max           outliers         delta
  wall_time           105s  ± 4.07s      101s  …  109s           0 ( 0%)        💩+ 24.5% ±  9.5%
  peak_rss           4.92GB ±  141KB    4.92GB … 4.92GB          0 ( 0%)        💩+  5.0% ±  2.4%
  cpu_cycles          410G  ± 4.33G      405G  …  414G           0 ( 0%)        💩+ 24.1% ±  3.8%
  instructions        596G  ±  200M      596G  …  596G           0 ( 0%)        💩+ 28.6% ±  4.4%
  cache_references   26.3G  ± 42.8M     26.2G  … 26.3G           0 ( 0%)        💩+ 10.0% ±  2.9%
  cache_misses       3.24G  ±  303M     3.05G  … 3.59G           0 ( 0%)        💩+ 81.8% ± 34.6%
  branch_misses      2.56G  ± 1.66M     2.56G  … 2.56G           0 ( 0%)        💩+ 14.1% ±  4.8%
```

*Increased* peak RSS? I'm not sure how that happened. Now I'm starting to doubt my methodology. Why are my results so different than both @jacobly0 and @antlilja?

**Comment by Jarred-Sumner** [comment]
i wonder if `toOwnedSlice` is increasing memory usage since it potentially re-allocates the shrunk array before freeing the old one? or, if arena allocator is in use when `toOwnedSlice` is called or on ArrayList that expand a lot that could also cause this

**Comment by antlilja** [comment]
> Well, I'm just glad we caught it now. Thanks for all your efforts on this branch.
> 
> Here's a new data point that includes #19069 vs old master:
> 
> ```
> Benchmark 1 (3 runs): 0.12.0-dev.2931+8d651f512/zig build-exe ...
>   measurement          mean ± σ            min … max           outliers         delta
>   wall_time          84.6s  ± 2.89s     81.3s  … 86.7s           0 ( 0%)        0%
>   peak_rss           4.68GB ± 70.2MB    4.64GB … 4.76GB          0 ( 0%)        0%
>   cpu_cycles          330G  ± 6.41G      327G  …  338G           0 ( 0%)        0%
>   instructions        464G  ± 12.6G      456G  …  478G           0 ( 0%)        0%
>   cache_references   23.9G  ±  427M     23.6G  … 24.4G           0 ( 0%)        0%
>   cache_misses       1.78G  ±  239M     1.51G  … 1.93G           0 ( 0%)        0%
>   branch_misses      2.25G  ± 66.6M     2.21G  … 2.32G           0 ( 0%)        0%
> Benchmark 2 (3 runs): master+19069/zig build-exe ...
>   measurement          mean ± σ            min … max           outliers         delta
>   wall_time           105s  ± 4.07s      101s  …  109s           0 ( 0%)        💩+ 24.5% ±  9.5%
>   peak_rss           4.92GB ±  141KB    4.92GB … 4.92GB          0 ( 0%)        💩+  5.0% ±  2.4%
>   cpu_cycles          410G  ± 4.33G      405G  …  414G           0 ( 0%)        💩+ 24.1% ±  3.8%
>   instructions        596G  ±  200M      596G  …  596G           0 ( 0%)        💩+ 28.6% ±  4.4%
>   cache_references   26.3G  ± 42.8M     26.2G  … 26.3G           0 ( 0%)        💩+ 10.0% ±  2.9%
>   cache_misses       3.24G  ±  303M     3.05G  … 3.59G           0 ( 0%)        💩+ 81.8% ± 34.6%
>   branch_misses      2.56G  ± 1.66M     2.56G  … 2.56G           0 ( 0%)        💩+ 14.1% ±  4.8%
> ```
> 
> _Increased_ peak RSS? I'm not sure how that happened. Now I'm starting to doubt my methodology. Why are my results so different than both @jacobly0 and @antlilja?

Hmm, maybe those results differ from @jacobly0 because he only does the bitcode emission without actually compiling to a binary? But maybe I'm misinterpreting this line:
> ... I compared the compiler compiling itself to unoptimized bitcode with an empty cache ...

I'll do some benchmarks as a sanity check

**Comment by andrewrk** [comment]
It's looking like, while the time and memory spent in Zig land is reduced, the time and memory spent in C++ land is increased even more. That's too bad. Looks like my prediction was very wrong.

That being said, it does not change the plan. This is a large step towards the long term goal of eliminating the library dependency on LLVM and providing fast compilation speed by avoiding LLVM altogether.

**Comment by andrewrk** [comment]
@Jarred-Sumner 
> i wonder if `toOwnedSlice` is increasing memory usage since it potentially re-allocates the shrunk array before freeing the old one? or, if arena allocator is in use when `toOwnedSlice` is called or on ArrayList that expand a lot that could also cause this

It's 100% guaranteed to reuse the same pointer when shrinking, when using the C allocator:

https://github.com/ziglang/zig/blob/31763d28c460933500916707954da1590bfe893d/lib/std/heap.zig#L121-L123

however I did notice a related possible improvement to make to `std.heap.raw_c_allocator`. Will PR shortly.

**Comment by antlilja** [comment]
> Another thought
> 
> What if LLVM's implementation uses the shorthand for repeated elements? Is the file size of the .bc files similar between the Zig-generated one and the LLVM-generated one?

I have some measurements in my new PR: #19083

File: src/codegen/llvm/Builder.zig [hunk 3]
```diff
@@ -11722,15 +11701,3260 @@ fn constantExtraData(self: *const Builder, comptime T: type, index: Constant.Ite
...
+) Allocator.Error!Metadata {
+    try self.ensureUnusedMetadataCapacity(1, Metadata.Subrange, 0);
+    return self.debugSubrangeAssumeCapacity(lower_bound, count);
+}
+
+pub fn debugExpression(
+    self: *Builder,
+    elements: []const u32,
+) Allocator.Error!Metadata {
+    try self.ensureUnusedMetadataCapacity(1, Metadata.Expression, elements.len * @sizeOf(u32));
```
Comment (jacobly0) [inline src/codegen/llvm/Builder.zig:12121]: Sorry for not noticing during the original review, but all of these calls are passing the wrong third argument.

Comment (antlilja) [reply src/codegen/llvm/Builder.zig:12121]: Submitted a fix: https://github.com/ziglang/zig/pull/20484

**Pull request merged**